scylla-proxy/src/proxy.rs

```diff
@@ -1665,4 +1680,56 @@ mod tests {
 
         running_proxy.finish().await.unwrap();
     }
+
+    #[tokio::test]
+    #[ntest::timeout(100)]
+    async fn sanity_check_reports_errors() {
+        let node1_real_addr = SocketAddr::from_str("127.0.0.58:9876").unwrap();
+        let node1_proxy_addr = SocketAddr::from_str("127.0.0.59:9876").unwrap();
+        let proxy = Proxy::new([Node::new(
+            node1_real_addr,
+            node1_proxy_addr,
+            ShardAwareness::Unaware,
+            None,
+            None,
+        )]);
+        let mut running_proxy = proxy.run().await.unwrap();
+
+        let mock_node_listener = TcpListener::bind(node1_real_addr).await.unwrap();
+
+        let mock_driver_action = async {
+            let mut conn = TcpStream::connect(node1_proxy_addr).await.unwrap();
+
+            conn.write_all(b"uselessJunk").await.unwrap();
+            conn
+        };
+
+        let mock_node_action = async {
+            let (conn, _) = mock_node_listener.accept().await.unwrap();
+            conn
+        };
+
+        let (node_conn, driver_conn) = join(mock_node_action, mock_driver_action).await;
+
+        running_proxy.sanity_check().unwrap();
+
+        mem::drop(driver_conn);
+        tokio::time::sleep(Duration::from_millis(10)).await;
+        assert_matches!(
+            running_proxy.sanity_check(),
+            Err(ProxyError::Worker(WorkerError::DriverDisconnected(_)))
+        );
```

Comment:
This assertion depends on timing, which is bad. The test should wait until an error occurs, not just sleep and then check.

Reply:
How can I achieve that, bearing in mind how `sanity_check()` works?

Reply:
You could have a version of `sanity_check` which waits for an error to occur in `error_sink`. E.g. split `sanity_check` into `check_for_error` and `wait_for_error`.